### PR TITLE
feat(medusa, medusa-js, medusa-react): Implement Sales Channel update

### PR DESCRIPTION
**Pull request by pKorsholm**
WIP - still missing medusa-js and medusa-react

**What**
- add update functionality for sales channel api

**Testing** 
- api tests: 
  - `returns updated sales channel`
  - `calls service update`
- service tests: 
  - `calls save with the updated sales channel`
  - `returns the saved sales channel` 
- integration test: 
  - `updates sales channel properties`

Fixes CORE-279

### CODE REVIEW

File: packages/medusa/src/services/sales-channel.ts [hunk 1]
```diff
@@ -54,7 +61,35 @@ class SalesChannelService extends TransactionBaseService<SalesChannelService> {
...
+        transactionManager.getCustomRepository(this.salesChannelRepository_)
+
+      const salesChannel = await salesChannelRepo.findOne(id)
+
+      if (!salesChannel) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `Sales Channel with ${id} was not found`
+        )
+      }
```
Comment (pKorsholm) [inline packages/medusa/src/services/sales-channel.ts:75]: replace with retrieve

**Review by adrien2p — COMMENTED** [review]
Looks nice! Just few suggestions 🎉

**Comment by olivermrbl** [comment]
Just closing and opening for Linear sync. to kick in @pKorsholm 👍

File: packages/medusa/src/api/routes/admin/sales-channels/update-sales-channel.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ *             type: string
+ *             description: Name of the sales channel.
+ *           description:
+ *             type: string
+ *             description:  Sales Channel description.
+ *           is_disabled:
+ *             type: boolean
+ *             description:  Indication of if the sales channel is active.
+ * tags:
+ *   - Customer
```
[the remark this answers is not in the document]
Comment (pKorsholm) [reply packages/medusa/src/api/routes/admin/sales-channels/update-sales-channel.ts]: To update the tag when you copy the oas, nice catch 😉

File: integration-tests/api/__tests__/admin/sales-channels.js [hunk 3]
```diff
@@ -37,6 +40,52 @@ describe("sales channels", () => {
...
+        payload,
+        {
+          headers: {
+            authorization: "Bearer test_token",
+          },
+        }
+      )
+
+      expect(response.status).toEqual(200)
+      expect(response.data.sales_channel).toMatchSnapshot({
```
[the remark this answers is not in the document]
Comment (pKorsholm) [reply integration-tests/api/__tests__/admin/sales-channels.js:128]: Good point, I have added it

File: packages/medusa/src/api/routes/admin/sales-channels/update-sales-channel.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/sales-channels/update-sales-channel.ts]: 😂 got it

File: packages/medusa/src/services/sales-channel.ts [hunk 4]
```diff
@@ -16,6 +17,12 @@ type InjectedDependencies = {
 }
 
 class SalesChannelService extends TransactionBaseService<SalesChannelService> {
+  static Events = {
+    CREATED: "sales_channel.created",
+    UPDATED: "sales_channel.updated",
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/services/sales-channel.ts:22]: Not sure we will avoid that, i have added the deleted in my pr 😂

File: integration-tests/api/factories/simple-sales-channel-factory.ts [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import { Connection } from "typeorm"
 import faker from "faker"
-import { SalesChannel } from "@medusajs/medusa"
+import { SalesChannel } from "@medusajs/medusa/dist/models/sales-channel"
```
Comment (olivermrbl) [inline integration-tests/api/factories/simple-sales-channel-factory.ts:3]: question: Should be left unchanged no?

File: packages/medusa/src/api/routes/admin/sales-channels/index.ts [hunk 6]
```diff
@@ -3,7 +3,9 @@ import { DeleteResponse, PaginatedResponse } from "../../../../types/common"
 import "reflect-metadata"
 import { isFeatureFlagEnabled } from "../../../middlewares/feature-flag-enabled"
 import { SalesChannel } from "../../../../models"
-import middlewares from "../../../middlewares"
+import middlewares, { transformBody } from "../../../middlewares"
+import "reflect-metadata"
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/sales-channels/index.ts:7]: comment: Would like to avoid this import, if possible

**Review by adrien2p — COMMENTED** [review]
LGTM! Few comments but nothing major 🚀

File: packages/medusa/src/services/sales-channel.ts [hunk 7]
```diff
@@ -54,7 +61,35 @@ class SalesChannelService extends TransactionBaseService<SalesChannelService> {
...
+      const salesChannel = await salesChannelRepo.findOne(id)
+
+      if (!salesChannel) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `Sales Channel with ${id} was not found`
+        )
+      }
+
+      for (const key of Object.keys(data).filter(
```
[the remark this answers is not in the document]
Comment (olivermrbl) [reply packages/medusa/src/services/sales-channel.ts:94]: suggestion(non-blocking): Really only to improve readibility and follow a pattern we've used elsewhere. Take it or leave it 😄 
```ts
for (const key of Object.keys(data) {
     if (typeof key !== "undefined") {
         salesChannel[key] = data[ley]
      }
}
```

**Review by olivermrbl — COMMENTED** [review]
Nice work! Left a couple of comments, but nothing big 👍

File: packages/medusa/src/api/routes/admin/sales-channels/index.ts (hunk 6 above)
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/sales-channels/index.ts:7]: Yes, why do we have that import here?

File: packages/medusa/src/services/sales-channel.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/services/sales-channel.ts:94]: Should we have an util that take an object and return an object without undefined value/key?

File: integration-tests/api/factories/simple-sales-channel-factory.ts (hunk 5 above)
Comment (adrien2p) [reply integration-tests/api/factories/simple-sales-channel-factory.ts:3]: Indeed, and if there is an error for the import it is most likely that we have to rebuild and re scan the deps

File: packages/medusa-js/src/resources/admin/sales-channels.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import {
   AdminSalesChannelRes,
```
[the remark this answers is not in the document]
Comment (pKorsholm) [reply packages/medusa-js/src/resources/admin/sales-channels.ts:2]: yes good catch 😄

**Review by olivermrbl — APPROVED** [review]
LGTM, just need a green pipeline 👍

**Pull request merged**